Added Release Workflow and Fixed some build script issues

jgmdev opened the pull request:
Ok, so this PR includes various changes regarding the build scripts and a new `release.yml` workflow in order to finally be able to hit the button and make a release that builds working packages for each platform.

The release workflow can be manually triggered and  it lets you enter a version so one can trigger a build and make a draft release which can then be adjusted:

![2022-06-01_14:41:54](https://user-images.githubusercontent.com/1702572/171478823-4a78eb8c-eb21-4ea4-af6d-d8523009c03b.png)

Should also work on pushed tags but haven't tested that, but if it doesn't works we can adjust later.

## Some of the Changes 

* make builds properly static by forcing the meson fallback
* we now only depend on meson and forcefallback to generate the release static builds
* install wayland protocols on linux target for proper wayland support
* disabled x11 and wayland under mac which resulted on additional dependencies
* CI builds are now debug builds so they include debug symbols and git commit on version
* enabled some video subsystems for linux so the appimage and portable build launch properly
* enabled use_timers which is required for proper sdl time calculation which we rely on (fixes statusview neve hiding messages)
* fixed msys detected platform to properly report i686 on mingw32 build run and not only x86_64
* copy missing dll's to portable windows zip file
* fixed inno setup copying from the bare bone data dir instead of the one generated by meson
* changed license which is used on innosetup to reflect the entire litexl team
* added settings gui as addon
* added language plugins as addon
* other changes and fixes...

Guldoman commented on scripts/common.sh:
```diff
@@ -2,6 +2,66 @@
+  local build_dir="$1"
+  local data_dir="$2"
+
+  for module_name in colors widget; do
```
`widget` should probably only installed if we get asked to add optional plugins (see next comment).

Guldoman commented on scripts/common.sh:
```diff
@@ -14,9 +74,23 @@ get_platform_name() {
+get_platform_arch() {
+  platform=$(get_platform_name)
+  arch=$(uname -m)
+  if [[ $MSYSTEM != "" ]]; then
```
Doesn't `uname -m` work in MSYS?

Guldoman commented on scripts/package.sh:
```diff
@@ -152,11 +131,15 @@ main() {
         fi
         shift
         ;;
+      -r|--release)
```
As above.

Guldoman commented on scripts/package.sh:
```diff
@@ -202,6 +186,14 @@ main() {
+        | grep mingw \
+        | awk '{print $3}' \
+        | sed 's#\\#/#g' \
+        | xargs -I '{}' cp -v '{}' "$(pwd)/${dest_dir}/"
```
Can't we statically link the dependencies?

Guldoman commented on scripts/appimage.sh:
```diff
@@ -16,22 +23,21 @@ show_help(){
+  echo "-s --static               Specify if building using static libraries."
   echo "-v --version VERSION      Specify a version, non whitespace separated string."
+  echo "-a --addons               Install 3rd party addons."
+  echo "-r --release              Compile in release mode."
```
As before, I'd prefer `release` by default.

Guldoman commented on build-packages.sh:
```diff
@@ -37,6 +37,7 @@ show_help() {
   echo "-D --dmg                  Create a DMG disk image (macOS only)."
   echo "                          Requires NPM and AppDMG."
   echo "-I --innosetup            Create an InnoSetup installer (Windows only)."
+  echo "-r --release              Compile in release mode."
```
I'd prefer it to be `release` by default, and `debug` with a switch.

Guldoman commented on scripts/package.sh:
```diff
@@ -252,9 +249,15 @@ main() {
-  if [[ $bundle == true && $dmg == true ]]; then source scripts/appdmg.sh "${package_name}"; fi
-  if [[ $innosetup == true ]]; then source scripts/innosetup/innosetup.sh -b "${build_dir}"; fi
+  if [[ $appimage == true ]]; then
+    source scripts/appimage.sh $flags --static
```
Does this make it build twice? Maybe we should specify `--nobuild`.

jgmdev replied:
The thing is that the first build makes some modifications to be able to execute lite-xl in a portable way, also the whole directory structure gets compacted into data, meanwhile an appimage is kind of a chroot with escalated privileges and the typical /usr/ directory tree is desired with regular lookup paths.

Guldoman replied:
Alright but this is a bit of a waste. Hopefully we can find a way to avoid doing this.

Guldoman commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,226 @@
+    inputs:
+      version:
+        description: Release Version
+        default: v2.1.0
```
Can this be deduced from the tag?

jgmdev replied:
That is just a place holder that gets pre-filled for you when manually triggering the workflow. look at the screenshot on first post, it could be replaced by vx.x.x

Guldoman commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,226 @@
+          tag_name: ${{ steps.tag.outputs.version }}
+          release_name: "Lite XL ${{ steps.tag.outputs.version }}"
+          draft: true
+          prerelease: false
```
Does it let you change to prerelease from a draft release?

jgmdev replied:
Yes, as mentioned on discord

Guldoman commented on meson.build:
```diff
@@ -108,23 +108,35 @@ if not get_option('source-only')
-    sdl_options += 'use_video_wayland=auto'
+    if host_machine.system() == 'darwin' or host_machine.system() == 'windows'
+        sdl_options += 'use_video_x11=disabled'
+        sdl_options += 'use_video_wayland=disabled'
```
I wonder if this is a wrap bug. I'd expect that those would be automatically disabled on unsupported platforms.

jgmdev replied:
MacOS supported X11 and there is still some xserver packages floating for it, also it seems that the github runner does have those packages installed so x11 support gets enabled and linked in, the issue is that the user most of the time will not normally have that library installed so we have to explicitly disable it at least for darwin, did it for windows too just in case...

Guldoman replied:
Fair enough.

Guldoman commented on meson.build:
```diff
@@ -108,23 +108,35 @@ if not get_option('source-only')
+        sdl_options += 'use_video_openglesv2=auto'
+    else
+        sdl_options += 'use_video_opengl=disabled'
+        sdl_options += 'use_video_openglesv2=disabled'
```
What happens if you leave them on `auto`?

jgmdev replied:
It gets enabled for the platform if support is found, but at least on windows there seem to be better backends that should perform better for that platform like DX11

Guldoman reviewed (commented):
My main gripes are:
* This basically duplicates `build.yml` in `release.yml`.
  Hopefully it's simple enough to integrate one in the other.
* The additional plugins packaged by default.
  Mainly the `settings` plugin which is a paradigm shift on how to customize the editor.
  Lite XL has always been customized in Lua and I believe that we should keep it that way.
  I suggest we release the normal "minimal" version and an "extended" version that includes additional plugins like `settings`.

Other than that, this should finally solve our problems with releases!

jgmdev commented on scripts/common.sh:
```diff
@@ -2,6 +2,66 @@
+  fi
+
+  # Download third party color themes
+  curl --insecure \
```
That wasn't added by me but I think I know why, the cause is that sometimes the workflow container/runner, lets say uubuntu-18.04... may not have updated trusted certificates causing the downloads to fail even when they are valid.

Guldoman replied:
Yeah but I feel like this poses a bit of a security problem. I wonder if the problem still happens.

jgmdev commented on scripts/common.sh:
```diff
@@ -2,6 +2,66 @@
+      "${data_dir}/plugins/"
+  done
+
+  rm "${build_dir}/third/data/plugins/language_cpp.lua"
```
I think I did, but I also think someone reintroduced it xD

Guldoman replied:
Let's remove it again then.

jgmdev replied:
done

jgmdev commented on .github/workflows/build.yml:
```diff
@@ -12,6 +12,7 @@ on:
   pull_request:
     branches:
     - '*'
+  workflow_dispatch:
```
Allows manually triggering the CI for testing and trouble shooting from the github `Actions` tab.

Guldoman replied:
How does it look like?
At the moment, without it, we can still re-trigger an Action by clicking on "Re-run all jobs". Does this enable something different?

Not-a-web-Developer replied:
it looks like this for the release action: 
<img width="266" alt="image" src="https://user-images.githubusercontent.com/47886897/173183148-f0409e01-9a3e-43ef-a8b4-e75b603efa4e.png">
the advantage of this is that it's only triggered when you tell it to, and it also asks for the version name before building which is nice; for the CI build, i agree that re-run all jobs is good enough.

Guldoman replied:
Yeah, but we're talking about the `build.yaml` one. Does it also add some new button?

Not-a-web-Developer replied:
build.yaml looks the same except for the prompt for release version, so yeah it's not exactly needed

jgmdev replied:
Something useful to have which doesn't hurts....
![2022-06-16_17:45:33](https://user-images.githubusercontent.com/1702572/174171391-73a8e260-e2e9-40ea-9b3c-a1e7f9f62d5b.png)

Guldoman replied:
Alright, so it *does* add a new button.

jgmdev commented on .github/workflows/build.yml:
```diff
@@ -54,7 +55,7 @@ jobs:
         echo "$HOME/.local/bin" >> "$GITHUB_PATH"
         echo "INSTALL_REF=${GITHUB_REF##*/}" >> "$GITHUB_ENV"
-        echo "INSTALL_NAME=lite-xl-${GITHUB_REF##*/}-linux-$(uname -m)" >> "$GITHUB_ENV"
+        echo "INSTALL_NAME=lite-xl-${GITHUB_REF##*/}-linux-$(uname -m)-portable" >> "$GITHUB_ENV"
```
Because that is the file name generated when using the --portable flag which last time I tested was needed to properly get the lite-xl linux packaged .tar to actually work...

Guldoman replied:
Our current releases don't add `-portable` to the name, so I don't think we need to rename this.

jgmdev commented on .github/workflows/build.yml:
```diff
@@ -65,21 +66,16 @@ jobs:
       run: |
         bash --version
-        bash scripts/build.sh --debug --forcefallback
+        bash scripts/build.sh --debug --forcefallback --portable
```
this tells meson that we want to generate installation files differently and lIRC properly set lookup paths for lite-xl, as I recall building without that flag resulted on lite-xl crashing on linux when starting up.

jgmdev commented:
> My main gripes are:
> 
>     * This basically duplicates `build.yml` in `release.yml`.
>       Hopefully it's simple enough to integrate one in the other.
> 
>     * The additional plugins packaged by default.
>       Mainly the `settings` plugin which is a paradigm shift on how to customize the editor.
>       Lite XL has always been customized in Lua and I believe that we should keep it that way.
>       I suggest we release the normal "minimal" version and an "extended" version that includes additional plugins like `settings`.
> 
> 
> Other than that, this should finally solve our problems with releases!

We can have it on a single file but I'm not a fan of Eye parsing YAML and pretty much prefer to have two workflow files for the different purposes which is much more clear to understand, If it was for my taste I would even clean build.yml to just include the CI build instructions and strip all comments/references of supporting releases on it. Also I introduced a flag `--release` that strips debugging symbols, because before the changes I worked on the CI, builds where done without debug symbols which should be useful for such kind of development builds. So release.yml AKA Release builds without debugging symbols and `--buildtype=release` while build.yml AKA CI builds with debugging symbols and `--buildtype=debug`

About the settings gui, we discussed this some time ago also with Adam and the conclusion was that the settings plugin would not be part of main repo but nothing against distributing as part of releases so that is what I'm doing here, also having the toolbarview open settings gui doesn't breaks the ability to open the user module and make your changes, it also doesn't breaks the open user module command, on the contrary, the settings gui even has a button to open it for you and kind of guides the user to it since most used options are now self discoverable:

![2022-06-08_20:07:16](https://user-images.githubusercontent.com/1702572/172737510-3f07f807-a0f5-4a3e-821c-b18a62570222.png)

Guldoman commented:
> We can have it on a single file but I'm not a fan of Eye parsing YAML and pretty much prefer to have two workflow files for the different purposes which is much more clear to understand, If it was for my taste I would even clean build.yml to just include the CI build instructions and strip all comments/references of supporting releases on it.

I don't understand what the problem is. The CI builds should basically be the same as the release builds, so the only difference would be a "Create release" step that gets executed only on tag pushes. There's really no need for different files.

> Also I introduced a flag `--release` that strips debugging symbols, because before the changes I worked on the CI, builds where done without debug symbols which should be useful for such kind of development builds.

Yes, but it would be better to leave the default behavior as `release` and only enable `debug` builds with a switch.
People that compile Lite XL by themselves now get slower debug builds for no reason.

> About the settings gui, we discussed this some time ago also with Adam and the conclusion was that the settings plugin would not be part of main repo but nothing against distributing as part of releases so that is what I'm doing here

Yes, but this is a big behavior change, that's why I was suggesting to keep two different releases, a "minimal" one with only the usual default plugins and another "extended" one with all the bells and whistles.

> also having the toolbarview open settings gui doesn't breaks the ability to open the user module and make your changes, it also doesn't breaks the open user module command, on the contrary, the settings gui even has a button to open it for you and kind of guides the user to it since most used options are now self discoverable

What once required a click, now requires 3.

About options discoverability, with the config specs we can now generate ad-hoc autocompletions for the user module.

jgmdev commented:
Updated this PR and as discussed above:

* now build.yml doesn't enables addons
* simplified build.yml since the release logic is now done on release.yml
* restored non addon builds for releases

Will squash merge this now since not doing any release after so many changes or having broken CI builds is getting silly.
